### ISSUE TITLE
lastversion returns None randomly

### ISSUE THREAD

**Issue by IlanBeloglovsky** [dvershinin/lastversion#44]
I'm using `lastversion` as python module in a code that being executed by celery
this is my code:
```python
try:
                version = lastversion.latest(f'{org}/{lib}', output_format='dict', pre_ok=False)

                if version:
                    # Use the version
                else:
                    logger.error(f"Failed to fetch version for {org}/{lib}, got: {str(version)}")
            except Exception as e:
                logger.exception(f"An error occurred while trying to fetch version for {org}/{lib}", e)
```

I execute this code inside a loop that has around 45 iteration
For some reason beyond my understanding from time to time the function returns `None` but when i look at my logs i see that the module in fact DID managed to get the last version of that packages that he then returns None for them:
![image](https://user-images.githubusercontent.com/75306021/105627083-35967180-5e3d-11eb-9cbd-21fbe1b1d51c.png)

It happens with `SiliconLabs/uC-OS2` (but not only)
When i run the same code in IDLE it works just fine

**Comment by dvershinin** [comment]
It is puzzling, can you try `pip install git+https://github.com/dvershinin/lastversion.git` (the latest commit) and add the new logs for when None is returned, please.

When a tag is being selected, it is being checked with GitHub "releases" API for extra checks, however there's no log of that for you...

**Comment by IlanBeloglovsky** [comment]
I've tried to change my `requirements.txt` to use `git+https://github.com/dvershinin/lastversion.git` but unfortunately i see no new logs in data dog

**Comment by dvershinin** [comment]
This needs more info… Can you make sure that GitHub API token is in use when running with Celery?
My only suspicious is that it's not being used / exceeded quota while trying to fetch formal release via API.
(normally it first tries to locate in the repo's feeds XML which doesn't use the API, but then It does a last check against the API to ensure it has no pre-release flag).

**Comment by IlanBeloglovsky** [comment]
In the logs I provided in the issue one of the logs says `Using API token` , is it enough to make sure the token is in fact being used?

**Comment by dvershinin** [comment]
Yes, that should be enough. Any more details about how/what code is being run between Celery vs IDLE?

**Comment by IlanBeloglovsky** [comment]
My coworker checked the celery on his local station, and it worked fine so we think the problem is not in celery as well.... 
the only difference is that the production code is run on k8s hosted on AWS

**Comment by IlanBeloglovsky** [comment]
I've dived into the code i tracked the following section:
```python
 if not ret or version > ret['version'] or tag_date > ret['tag_date']:
                    # we always want to return formal release if it exists, cause it has useful
                    # data grab formal release via APi to check for pre-release mark
                    r = self.repo_query('/releases/tags/{}'.format(tag_name))
                    if r.status_code == 200:
                        formal_release = r.json()
                        if not pre_ok and formal_release['prerelease']:
                            log.info(
                                "Found formal release for this tag which is unwanted "
                                "pre-release: {}.".format(version))
                            continue
                        if ret and tag_date + timedelta(days=365) < ret['tag_date']:
                            log.info('The version {} is newer, but is too old!'.format(version))
                            break
                        # use full release info
                        ret = formal_release
                        ret['tag_name'] = tag_name
                        ret['tag_date'] = tag_date
                        ret['version'] = version
                        ret['type'] = 'graphql'
                    else:
                        if ret and tag_date + timedelta(days=365) < ret['tag_date']:
                            log.info('The version {} is newer, but is too old!'.format(version))
                            break
                        ret = {
                            'tag_name': tag_name,
                            'tag_date': tag_date,
                            'version': version,
                            'type': 'graphql'
                        }```
i have a reason to believe that we the condition on the if is not met for some reason because i can track the logs before it but if the if the condition were met than ret would have to have a value

**Comment by dvershinin** [comment]
The parts you're referring to are for GraphQL query. But repositories with few tags will be early satisfied by the previous mechanism which is [atom feeds](https://github.com/dvershinin/lastversion/blob/master/lastversion/GitHubRepoSession.py#L384-L434) parsing. E.g. with the newly added logging:

```
lastversion SiliconLabs/uC-OS2 --verbose
lastversion.lastversion - INFO - Verbose output.
lastversion.lastversion - INFO - Using cache directory: /Users/dvershinin/Library/Caches/lastversion.
lastversion.ProjectHolder - INFO - Created instance of GitHubRepoSession
lastversion.GitHubRepoSession - INFO - Using API token.
lastversion.GitHubRepoSession - INFO - Got HTTP status code 200 from https://github.com/SiliconLabs/uC-OS2/releases.atom
lastversion.GitHubRepoSession - INFO - Checking tag v2.93.00
lastversion.ProjectHolder - INFO - Sanitizing string v2.93.00 as a satisfying version.
lastversion.ProjectHolder - INFO - Parsed as Version OK
lastversion.ProjectHolder - INFO - String representation of version is 2.93.0.
lastversion.GitHubRepoSession - INFO - Got HTTP status code 404 from https://api.github.com/repos/SiliconLabs/uC-OS2/releases/tags/v2.93.00
lastversion.GitHubRepoSession - INFO - No formal release for tag v2.93.00
lastversion.GitHubRepoSession - INFO - Selected version as current selection: 2.93.0.
lastversion.lastversion - INFO - Located the latest release at: https://github.com/SiliconLabs/uC-OS2 via feed mechanism
2.93.0
```

The graphql is used when there are many tags in repository and using a "major" filtering - in that case going through feed (which contains only a subset) of all repo tags, will not yield desired result. So it uses graphql because it needs to look at "all tags":

```
lastversion php:7.2 --verbose
...
lastversion.lastversion - INFO - Located the latest release at: https://github.com/php/php-src via graphql mechanism
7.2.34
```

In short: The lastversion code should not reach far enough to graphql code for a single-tag repo `SiliconLabs/uC-OS2`.
If it did reach graphql code for you, then somehow it failed to parse this feed: https://github.com/weston-embedded/uC-OS2/releases.atom

I've added more logging for failure parse the feed or when no elements are present in the parsed feed.

**Comment by IlanBeloglovsky** [comment]
> lastversion.GitHubRepoSession - INFO - Got HTTP status code 200 from https://github.com/SiliconLabs/uC-OS2/releases.atom
> lastversion.GitHubRepoSession - INFO - Checking tag v2.93.00

I dont see these 2 logs in my logs, but my log level is set to INFO

**Comment by dvershinin** [comment]
Those are in master branch. You need to pip install from GitHub url or master tarball/zip

**Comment by IlanBeloglovsky** [comment]
ok i executed the code from the pode using python cli and finally i have some more informative logs(pip for some reason didnt download the latest commit of the repo)!
```
Using cache directory: /root/.cache/lastversion.
Created instance of GitHubRepoSession
Using API token.
Got HTTP status code 404 from https://github.com/SiliconLabs/uC-OS2/releases.atom
Failed to parse feed: syntax error
Feed has no elements. Empty repo???
Got HTTP status code 404 from https://api.github.com/repos/SiliconLabs/uC-OS2/releases/latest
Sanitizing string v2.93.00 as a satisfying version.
Parsed as Version OK
String representation of version is 2.93.0.
no tagger = no commit date, very old/imported tag isn't what
```

**Comment by dvershinin** [comment]
So that was all useful :)

Note how the repo name is actually now `weston-embedded/uC-OS2` and not `SiliconLabs/uC-OS2` which you're using.
Normally there is no issue with it: GitHub always redirects to the new repo name in both the API, and any link in repository web UI (e.g. releases.atom is redirected).

But for whatever reason, GitHub is not giving you the redirect for releases.atom when requested from AWS. And gives just the 404. This fails the first and most efficient method. It's an actual issue with GitHub servers not behaving consistently.

Now, the workaround I added is if it sees a 404 for the releases.atom, it will try to use the API (and hope that the redirect does happen there), to discover the new repo name and re-fetch the feed from the proper name URL.

The graphql issue is actual. The "tagger" exists only for annotated tags, and the other items returned by query which don't have it represent a non-annotated tag. I mistakenly thought we shouldn't bother with those and such only exist in old imported code (from the CVS), but it's pretty common. Replaced with what seems to be a more reliable query found [here](https://stackoverflow.com/questions/53862730/github-list-of-tag-and-release-asset-urls).
So the new logic is that we use actual tag (annotated) whenever available, and fallback to commit's (non-annotated tag) date.

Please test again using latest commit, and if it's all good will tag a new version.

**Comment by IlanBeloglovsky** [comment]
i run the tool straight from the cli and thats the error i got:
```
Traceback (most recent call last):
  File "/usr/local/bin/lastversion", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/site-packages/lastversion/lastversion.py", line 324, in main
    args.shorter_urls, args.major, args.only, args.at)
  File "/usr/local/lib/python3.6/site-packages/lastversion/lastversion.py", line 78, in latest
    release = s.get_latest(pre_ok=pre_ok, major=major)
  File "/usr/local/lib/python3.6/site-packages/lastversion/GitHubRepoSession.py", line 531, in get_latest
    ret = self.find_in_tags_via_graphql(ret, pre_ok, major)
  File "/usr/local/lib/python3.6/site-packages/lastversion/GitHubRepoSession.py", line 278, in find_in_tags_via_graphql
    if not j['data']['repository']['tags']['edges']:
TypeError: 'NoneType' object is not subscriptable
```

**Comment by dvershinin** [comment]
> C:\>lastversion SiliconLabs/uC-OS

@IlanBeloglovsky it just looks like you forgot trailing `2` in `SiliconLabs/uC-OS2`, so that's a project that doesn't exist.
Added check so it will raise BadProjectError for that case.

**Comment by IlanBeloglovsky** [comment]
> > C:>lastversion SiliconLabs/uC-OS
> 
> @IlanBeloglovsky it just looks like you forgot trailing `2` in `SiliconLabs/uC-OS2`, so that's a project that doesn't exist.
> Added check so it will raise BadProjectError for that case.

You are right, i forgot the trailing 2 when i run it localy, but not when i ran it from aws, i found the problem tho. opening a PR

**Comment by IlanBeloglovsky** [comment]
> > C:>lastversion SiliconLabs/uC-OS
> 
> @IlanBeloglovsky it just looks like you forgot trailing `2` in `SiliconLabs/uC-OS2`, so that's a project that doesn't exist.
> Added check so it will raise BadProjectError for that case.

You're right about the error when i ran it locally, but it still wont work from aws. 
```python
import requests

url = "https://api.github.com/graphql"

payload="""
        {
          rateLimit {
            cost
            remaining
          }
          repository(owner: "SiliconLabs", name: "uC-OS2") {
            tags: refs(refPrefix: "refs/tags/", first: 100, after: "", 
              orderBy: {field: TAG_COMMIT_DATE, direction: DESC}) {
              edges {
                cursor,
                node {
                  ...refInfo
                }
              }
            }
          }
        }
        
        fragment refInfo on Ref {
          name
          target {
            sha: oid
            commitResourcePath
            __typename
            ... on Tag {
              target {
                ... on Commit {
                  ...commitInfo
                }
              }
              tagger {
                name
                email
                date
              }
            }
            ... on Commit {
              ...commitInfo
            }
          }
        }
        
        fragment commitInfo on Commit {
          zipballUrl
          tarballUrl
          author {
            name
            email
            date
          }
        }
        
        
        """
headers = {
  'Authorization': 'token 750c17dac2b15d852cb8d50f8e871a7dac5981fd',
  'Content-type': 'application/json'
}

response = requests.request("POST", url, headers=headers, data=payload)

print(response.text) # {"message":"Problems parsing JSON","documentation_url":"https://docs.github.com/graphql"}
```
